src/binder/activitytracker.cpp

```diff
@@ -0,0 +1,80 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+// ============================================================
+//
+// activitytracker.cpp
+//
+
+
+//
+// Helpers for interaction with the managed ActivityTracker
+//
+// ============================================================
+
+#include "common.h"
+#include "activitytracker.h"
+
+void ActivityTracker::Start(const WCHAR *providerName, const WCHAR *activityName, /*out*/ GUID *activityId, /*out*/ GUID *relatedActivityId)
+{
+    GCX_COOP();
+    struct _gc {
+        OBJECTREF activityTracker;
+        STRINGREF providerName;
+        STRINGREF activityName;
+    } gc;
+    ZeroMemory(&gc, sizeof(gc));
+
+    GCPROTECT_BEGIN(gc);
+
+    MethodDescCallSite getInstance(METHOD__ACTIVITY_TRACKER__GET_INSTANCE);
```

Comment:
It is best to minimize amount of "manually managed" code and push as much as possible to the managed side. In this case, it would mean:

- Change the managed method to take pointers and create the managed strings from pointers on the managed side
- Change the managed method to be static and make it do both GetInstance and OnStart.

Once you do that you won't need the GCPROTECT and you need just a single MethodDescCallSite.

Reply:
Yay! Switched to static methods for this (and using a managed string constant)